Wedderburn

piotrmizerka opened the pull request:
@kalmarek I implemented a matrix example as described in my notes: https://github.com/piotrmizerka/postdoc-warsaw-daily-notes/blob/main/pdf/2022_07_28_matrix_Wedderburn.pdf . Now, I think again that my reasoning is correct there (I described the definition of the action of S_3 in line 115 in "experiment_wedderburn.jl". A more detailed explanation below:

I think one has to distinguish between two situations

(1) two actions on R^n and R^m defined by reps rho: G-->GL_n(R), phi: H-->GL_m(R) define the action of GxH on the tensored space R^n\otimes R^m\cong R^{nm} given by (g,h)\mapsto rho(g)\otimes phi(h). I think this is the action you have in mind (in your e-mail from 13 August) in your comment that rho_0 is the representation of S_2x\Sigma.

(2) suppose we have two actions of the same group G, given by reps rho: G-->GL_n(R), phi: H-->GL_m(R). Then we can define the action of G on R^n\otimes R^m\cong R^{nm} by (g,h)\mapsto rho(g)\otimes phi(g).

One can therefore think of two actions: one broader, of GxG on R^{nm} (special case of (1)) and the second (the one we desire I suppose, the case (2)) of G on R^{nm}. 

In your message, you also wrote that one should define the action of \Sigma on {1,2} by pi_3 - but this action would be faithful, whereas ours is not...

Summing up: I think for our purpose (G=<s_1,...,s_n|...>) one shall consider the basis i\otimes e, where i=1,...,n and e\in E (E - e.g. half basis) and define the actions of a subgroup G of S_n by permutations on {1,...,n} and E respectively. These actions define the action on our basis {i\otimes e} (note that the basis is of size n|E|).  So in the example I provided we took G=\Sigma=S_3 acting on E={x,y,z} by permutations, and on {1,2} as describe in line 115. Typically, the set E will be much bigger than the order of permutations we consider.

piotrmizerka committed:
basic Wedderburn example from my notes

piotrmizerka committed:
symmetrized matrix reconstruction

caveat: unregistered PropertyT-new added - remember to add it yourself

piotrmizerka committed:
matrix example

TODO: now, the action is defined in a wrong way

piotrmizerka committed:
surround headings by blank lines

piotrmizerka committed:
round matrix entries for better visibility

piotrmizerka committed:
matrix example

piotrmizerka committed:
remove PropertyT_new

piotrmizerka committed:
a reference comment placed more adequately

piotrmizerka committed:
PropertyT-new temporarily added

piotrmizerka committed:
Wedderburn setting - now working yet

piotrmizerka committed:
remove PropertyT-new as it is not registered

piotrmizerka committed:
Wedderburn desomposition for SAutFn

Caution: implementation requires polishing up! Verification required!

kalmarek commented:
this is what we came up today:
`/src/alphabet_permutation.jl`
```julia
struct AlphabetPermutation{GEl,I} <: SymbolicWedderburn.ByPermutations
    perms::Dict{GEl,Perm{I}}
end
function AlphabetPermutation(
    A::Alphabet,
    Γ::PermutationGroups.AbstractPermutationGroup,
    op,
)
    return AlphabetPermutation(
        Dict(γ => inv(Perm([A[op(l, γ)] for l in A])) for γ in Γ),
    )
end
function AlphabetPermutation(A::Alphabet, W::Constructions.WreathProduct, op)
    return AlphabetPermutation(
        Dict(
            w => inv(Perm([A[op(op(l, w.n), w.p)] for l in A])) for
            w in W
        ),
    )
end

function Base.:^(
    w::Groups.AbstractWord,
    p::PermutationGroups.AbstractPerm,
)
    return typeof(w)([l^p for l in w])
end
function SymbolicWedderburn.action(
    act::AlphabetPermutation,
    γ::GroupElement,
    g::Groups.AbstractFPGroupElement,
)
    G = parent(g)
    w = word(g)^(act.perms[γ])
    return G(w)
end
```

defined somewhere locally:
```
function _conj(
    t::Groups.Transvection,
    σ::PermutationGroups.AbstractPerm,
)
    return Groups.Transvection(t.id, t.i^σ, t.j^σ, t.inv)
end

function _conj(
    t::Groups.Transvection,
    x::Groups.Constructions.WreathProductElem,
)
    tσ = _conj(t, x.perm)
    dual_id = ifelse(t.id == :ϱ, :λ, :ϱ)
    dual_inv = ifelse(t.inv, false, true)
    new_id = isone(x.elts[t.i]) ? t.id : dual_id
    new_inv = isone(x.elts[t.j]) ? t.inv : dual_inv

    return Groups.Transvection(new_id, tσ.i, tσ.j, new_inv)
end

n = 3
G = SAut(FreeGroup(n))
Σ = WreathProduct(SymmetricGroup(2), SymmetricGroup(n))
act = AlphabetPermutation(alphabet(G), Σ, _conj)
```

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+    row_generator::Groups.GroupElement
+    column_generator::Groups.GroupElement
+    entry::Groups.GroupElement
+end
```
we could write this as
```julia
struct TensorSupportElement{B<:AbstractBasis}
    i::Int
    j::Int
    g::Int
    basis::B
end
```

kalmarek replied:
```julia
struct TensorSupportElement{GEl}
     i::GEl
     j::GEl
     k::GEl
end
```
gives much clearer code

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+        result = result*act_on_transvection(σ, SAutFn(word[i:i]))
+    end
+    return result
+end
```
replaced by `SymbolicWedderburn.action(act, σ, g)`

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+        wreath_conjugation(g, tensor_support_element.column_generator),
+        wreath_conjugation(g, tensor_support_element.entry)
+    )
+end
```
```julia
function SymbolicWedderburn.action(
    act::AlphabetPermutation,
    g::Groups.GroupElement
    tse::TensorSupportElement,
)
    i = SymbolicWedderburn.action(act, tse.i, g)
    j = SymbolicWedderburn.action(act, tse.j, g)
    k = SymbolicWedderburn.action(act, tse.g, g)
    return TensorSupportElement(i, j, k)
end
```

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+    end
+
+    return constraints_basis, psd_basis
+end
```
```julia
function matrix_bases(basis, half_basis, S)
     constr_basis = [TensorSupportElement(s, t, g) for s in S for t in S for g in basis]    
     psd_basis = [TensorSupportElement(s, s, g) for s in S for g in half_basis]
    return constr_basis, psd_basis
end
```

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+end
+
+function wedderburn_decomposition_matrix(
+    Σ,
```
```suggestion
    action,
    Σ,
```

piotrmizerka replied:
we don't need this function to be defined separately

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,194 @@
+    half_basis,
+    S # stands for the generating set we choose
+)
+    action = MatrixAction()
```
define externally as `action = AlphabetPermutation(A, Σ, _conj)`

piotrmizerka committed:
Wedderburn for SAutFn polished up

piotrmizerka committed:
check order of automorphism composition

piotrmizerka committed:
changed for left-action convention

piotrmizerka committed:
don't repeat comments

piotrmizerka committed:
integrate Wedderburn functions with the module

piotrmizerka committed:
remove repeated code and apply left actions

piotrmizerka committed:
add wedderburn.jl to the module

piotrmizerka committed:
tests for Wedderburn

kalmarek commented on src/wedderburn.jl:
```diff
@@ -76,3 +76,16 @@ basis, sizes = Groups.wlmetric_ball(S, radius = 2*half_radius)
+evaluate(elt) == (s1*s2, s2*s1*s2) # first right then left
+evaluate(elt) == (s1*s2*s1, s2*s1) # first left then right
+# It turns out that the convention is the classical one: compose automorphisms first
+# from the right then from the left.
```
@piotrmizerka 
```julia
julia> saut = SpecialAutomorphismGroup(FreeGroup(3))
automorphism group of free group on 3 generators

julia> S = gens(saut)
12-element Vector{Automorphism{FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}, …}}:
 ϱ₁.₂
 ϱ₁.₃
 ϱ₂.₁
 ϱ₂.₃
 ϱ₃.₁
 ϱ₃.₂
 λ₁.₂
 λ₁.₃
 λ₂.₁
 λ₂.₃
 λ₃.₁
 λ₃.₂

julia> g = S[1]*S[2]
 ┌ ϱ₁.₂*ϱ₁.₃:
 │   f1 ↦ f1*f2*f3
 │   f2 ↦ f2
 └   f3 ↦ f3
```
so `g` acts on `domain(g) = (f1, f2, f3)` as follows
 * we first applied `S[1]=ϱ₁.₂` to `(f1, f2, f3)` obtaining `(f1·f2, f2, f3)`
 * we then applied `S[2]=ϱ₁.₃` to `((f1·f2), f2, f3)` obtaining `((f1·f2)·f3, f2, f3)`
as far as I understand this is action **on the right**. Am I mistaken?

piotrmizerka replied:
I think, I understand the composition differently here... My comprehension of how to apply `S[2]=\rho_1,3` to `f1*f2`: 

`f1*f2` goes to `S[2](f1)*S[2](f2)=f1*f3*f2`.

Hence, assuming action on the right, `(S[1]*S[2])=(f1*f3*f2, f2, f3)`.

On the other hand, if we assume action on the left, then:

1. we apply first `S[2]`: `S[2](f1, f2, f3) = (f1*f3, f2, f3)`,
2. and then we apply `S[1]`: `(S[1]*S[2])(f1, f2, f3) = S[1](S[2](f1, f2, f3)) = S[1](f1*f3, f2, f3) = (f1*f2*f3, f2, f3)`.

piotrmizerka committed:
a bug with cyclic group of order 2

piotrmizerka committed:
symmetrized matriz sdp problem

piotrmizerka committed:
averaged constraints do not have to be invariant!

piotrmizerka committed:
still, the constraint matrix has to be invariant

bugs with half_basis fixed

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -0,0 +1,87 @@
+    cnstrs = constraints(A.mstructure)
+    @assert length(cnstrs) == length(basis(A))
+
+    constraints_basis = w_dec_matrix.basis
+    considered_constraints = Dict{eltype(constraints_basis), Bool}()
+    for constraint in constraints_basis
+        considered_constraints[constraint] = false
+    end 
```
constraints_basis = SymbolicWedderburn.basis(w_dec_matrix)
```suggestion
    constraints_basis = SymbolicWedderburn.basis(w_dec_matrix)
    considered_constraints = Dict(c=>false for c in constraints_basis)
```
Why do we need to store all of these in a dict?

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -0,0 +1,87 @@
+    action = AlphabetPermutation(alphabet(parent(first(S))), Σ, _conj)
+    A_gs = Dict(g => A_g for (A_g, g) in zip(cnstrs, basis(A)))
+
+    for constraint in constraints_basis
```
shouldn't this loop go over the invariant vectors of `w_dec_matrix`?

piotrmizerka replied:
Yeah. That would be more reasonable... As for now, the version with the iteration over the whole constraint basis is left. I think that the total number of iterations would be O(constraint_basis size)). I guess it is hard to beat this when iterating over all invariant vectors as each invariant vector has length = constraint_basis size and one has to iterate over it somehow...

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -0,0 +1,87 @@
+            rhs = 0
+            for (π, ds) in pairs(SymbolicWedderburn.direct_summands(w_dec_matrix))
+                Uπ = SymbolicWedderburn.image_basis(ds)
+                rhs += dot(degree(ds)*Uπ*δ_i_j_g_invariant*Uπ', P[π])
```
```suggestion
                rhs += degree(ds)*dot(Uπ*δ_i_j_g_invariant*Uπ', P[π])
```
algebra is associative, but storage isn't ;)

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -0,0 +1,87 @@
+            i, j = word(constraint.row_generator)[1], word(constraint.column_generator)[1]
+            lhs = M[i,j](constraint.entry)
+            δ_i_j_g_invariant = invariant_constraint_matrix(constraint, Σ, action, psd_basis, A_gs, length(S))
+            rhs = 0
```
what is the type of `rhs`?

piotrmizerka replied:
Instead of `0` one could write `zero(typeof(lhs))`, I guess...

piotrmizerka committed:
more appropriate test name for symmetrization

piotrmizerka committed:
relations require more general type

piotrmizerka committed:
reconstructed invariant P incorporated into module

piotrmizerka committed:
SAutFn script

CAUTION: this is a preliminary, indexified, and not working version :)

piotrmizerka committed:
compute Wedderburn dec from matrix bases

piotrmizerka committed:
SAut_F_4 data filled

piotrmizerka committed:
remove @infos

piotrmizerka committed:
fix: add gens to basis when computing Fox ders

piotrmizerka committed:
tensor support element redefined more concisely

piotrmizerka committed:
we don't need a seperate wrap-up for Wedderburn

piotrmizerka committed:
no need for an additional def of _conj

piotrmizerka committed:
tensor support elt type test customized

piotrmizerka committed:
invariant matrix's test failure hopefully outdated

piotrmizerka committed:
no need to store whole constraint basis in a dict

piotrmizerka committed:
action on constraints redefined

a separate action of wreath product on group elts added

piotrmizerka committed:
matrix_bases function in a concise manner

piotrmizerka committed:
Wedderburn-related code in a single file

piotrmizerka committed:
debugging checkpoints added

piotrmizerka committed:
SAutFn: typos in relations corrected

kalmarek commented on src/wedderburn.jl:
```diff
@@ -0,0 +1,76 @@
+function AlphabetPermutation(A::Alphabet, G::Groups.Constructions.WreathProduct, op)
+    return AlphabetPermutation(
+        Dict(
+            g => PermutationGroups.Perm([A[op(l, g)] for l in A.letters]) for
```
`A.letters` → `A`

piotrmizerka committed:
right action convention

kalmarek commented on scripts/SAut_F_4.jl:
```diff
@@ -0,0 +1,168 @@
+
+function determine_transvection(
+    g::Groups.GroupElement # has to be a generating transvection or its inverse
+)
+    SAutFn = parent(g)
+    Fn = SAutFn.group
+    Fn_gens = Groups.gens(Fn)
+    g_id, g_i, g_j, g_inv = 0, 0, 0, 0
+    for k in eachindex(Fn_gens)
+        if g(Fn_gens[k]) != Fn_gens[k]
+            g_i = k
+            for l in eachindex(Fn_gens)
+                if g(Fn_gens[k]) == Fn_gens[k]*Fn_gens[l]
+                    g_id = :ϱ
+                    g_j = l
+                    g_inv = false
+                    break
+                elseif g(Fn_gens[k]) == Fn_gens[k]*Fn_gens[l]^(-1)
+                    g_id = :ϱ
+                    g_j = l
+                    g_inv = true
+                    break
+                elseif g(Fn_gens[k]) == Fn_gens[l]*Fn_gens[k]
+                    g_id = :λ
+                    g_j = l
+                    g_inv = false
+                    break
+                elseif g(Fn_gens[k]) == Fn_gens[l]^(-1)*Fn_gens[k]
+                    g_id = :λ
+                    g_j = l
+                    g_inv = true
+                    break
+                end
+            end
+            break    
+        end
+    end
+
+    return Groups.Transvection(g_id, g_i, g_j, g_inv)
+end
```
`Alphabet` defines a bijection between 1:n and letters of alphabet, ie.
```
A = Alphabet(...)
l = A[1] # → the first letter of A
A[l] # the index of l, i.e. 1
```
Actually it is also possible to do `A[-1]` to obtain __the inverse__ of the first letter (if it exists).

```julia
function determine_transvection(g::Automorphism)
    @assert parent(g) isa Automorphism{<:FreeGroup}
    @assert length(word(g)) = 1
    A = alphabet(parent(g))
    return A[first(word(g))]
end

kalmarek replied:
hmm? why resolved when the old `determine_transvection` is still there?

piotrmizerka replied:
sorry, forgot to change it, now shall be ok ;)

piotrmizerka committed:
arbitrary generating set in Fox calculus

piotrmizerka commented:
Coming back to the case when we symmetrize SAutFn with the wreath product. Suppose SAutFn has N generators. Still, I think we have to work out Fox derivatives with F_2N instead of F_N. In the Lyndon's paper, https://www.jstor.org/stable/1969440#metadata_info_tab_contents, he works strictly with the presentation G=F/R, where F is a free group. I think, this setup is quite strict. If we would work with F_N instead of F_2N, we would not have such a situation. For example, suppose x is a generator of SAutFn which we first treat as a generator of F_2N. the inverse of x in SAutFn corresponds to another generator, say y of F_2N, of whom we do know nothing a priori (unless we introduce a specific relation, which we do in a moment). Indeed, the relation we introduce is x*y=1, its Fox derivative over x, d(x*y)/dx equals 1. On the other hand, if we pass right away to F_N, then y is just the inverse of x in F_N. Therefore, in F_N, d(x*y)/dx = d(1)/dx = 0.

kalmarek commented:
Hmm... so let's do a concrete example:
* `SAut(F₄)` is generated (as a group!) by 24 elements (12 `ϱ`s and 12 `λ`s)
* (per Lyndon Lemma 5.1) to obtain `²M¹` we need to take `∂Rᵢ/∂sⱼ` for each relation `Rᵢ` defining `SAut(F₄)` and each of 24 generators `sⱼ` of it.

So at least in principle we don't need to take derivatives wrt. the inverses of generators. But somehow we do in here: https://github.com/piotrmizerka/LowCohomologySOS/blob/28ab152fac8e950b107f0475ad8914d4618ff525/scripts/SAut_F_4.jl#L67-L77

Why was it necessary? Per Lyndon paper we don't also need to take derivatives of the relations that are present in the free group, so the example you gave seems not applicable? hmm...

piotrmizerka committed:
determine_transvection compactified

piotrmizerka commented:
> 

Yes, we don't need to take inverses in order to compute cohomology from the matrices Mi. However, we need inverses for Wedderburn: we want to act on M_n(RG) by the wreath product (which flips the generators among others). That's why we need to add the inverses to the generating set. Now, as we have the symmetrized generating set, according to Lyndon, we work with the free group generated by this symmetrized set - this is F_48. So, in order to obtain SAutF_4 as a quotient, we have to introduce several new relations, e.g. those which tell that the gens corresponding to inverses are the indeed inverses.

piotrmizerka committed:
invariant constraints through Wedderburn data

piotrmizerka committed:
upgrade to Julia 1.8

piotrmizerka committed:
tiny changes

kalmarek commented:
Ok, I understand. 

We probably need to asses if using the wreath product action actually leads to smaller problem (we start with something larger, so we get more symmetries) than using just the symmetric group (we start with something smaller, so we get smaller symmetries).
Usually there's not much sense in first symmetrizing the problem only to use symmetry to reduce its size ;)

But maybe here we gain something more?

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -1,55 +1,38 @@
-        A_σg = A_gs[σ_orbit_representative.k]
-        result += δ_σi_σj⊗A_σg
+    bs = length(A_gs)
+    n = floor(Int, sqrt(div(length(v_inv), bs)))
```
`isqrt`

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -72,78 +55,33 @@ function sos_problem_symmetrized(
-        # if it%100 === 0
-        #     @info it
+    m = size(first(cnstrs))[1] # this is equal to the half_basis' size
+    n = floor(Int, sqrt(div(length(basis(w_dec_matrix)), length(basis(A)))))
```
`isqrt`

kalmarek reviewed (commented):
the symmetrized `sos_problem` looks good!

kalmarek commented on .github/workflows/CI.yml:
```diff
@@ -9,7 +9,7 @@ jobs:
       matrix:
         version:
-          - '1.6'
+          - '1.8'
```
why are we no longer testing on julia `1.6`?

piotrmizerka committed:
speed up symmetrized sos problem definition

piotrmizerka committed:
SAutF_4 script

piotrmizerka committed:
customize certification for Wedderburn

piotrmizerka committed:
symmetrized positive approx tests prettied up

piotrmizerka committed:
keep Julia 1.6 in tests

piotrmizerka committed:
isqrts

piotrmizerka committed:
change PropertyT_new to PropertyT

piotrmizerka committed:
don't print to much for larger cases

piotrmizerka committed:
JuMP referenced neatly

piotrmizerka committed:
SAut_Fn script prettied up (with deindexification)

piotrmizerka committed:
@time descr cause problems (at least in Julia 1.7)

piotrmizerka committed:
4 byte basis elts for Wedderburn

piotrmizerka committed:
4 byte basis elts for Wedderburn

piotrmizerka committed:
Merge branch 'Wedderburn' of https://github.com/piotrmizerka/LowCohomologySOS into Wedderburn

kalmarek commented on src/positive_approx_symmetrized.jl:
```diff
@@ -1,5 +1,5 @@
 basis(A::StarAlgebras.StarAlgebra) = A.basis
-basis(w::WedderburnDecomposition) = w.basis
+basis(w::SymbolicWedderburn.WedderburnDecomposition) = w.basis
```
why not just importing `StarAlgebras.basis` ?

piotrmizerka committed:
wedderburn action as a subtype of ByPermutations

piotrmizerka committed:
wedderburn action as a subtype of ByPermutations

piotrmizerka committed:
4-byte Wedderburn: corrected

piotrmizerka committed:
4-byte Wedderburn: corrected

piotrmizerka committed:
typo corrected

piotrmizerka committed:
typo corrected

piotrmizerka committed:
temporary solution to exceed UInt16

piotrmizerka committed:
temporary solution to exceed UInt16

kalmarek commented on src/group_rings.jl:
```diff
@@ -5,6 +5,12 @@ Base.copy(X::AlgebraElement) =
+        throw("index type $I is to small for basis of length $(length(basis))")
+    # @assert !(eltype(basis) <: Integer)
+    return StarAlgebras.Basis(basis, Dict(b => I(idx) for (idx, b) in pairs(basis)))
+end
```
hey, hey, hey @piotrmizerka :D the assertion is there because it's needed, not because of a whim :D

`Basis` is a vector which also implements a bijection between indices of `basis` vector and its elements. You can't distinguish between the two if both are integers: if `basis = [2,1]` then is `basis[1]` the first element or the index of `1` in `basis`?

kalmarek commented on src/wedderburn.jl:
```diff
@@ -51,9 +51,9 @@ function subset_permutation(
     act::AlphabetPermutation
 )
-    subset_idies = Dict(subset[i] => i for i in Int32.(eachindex(subset)))
+    subset_idies = Dict(subset[i] => i for i in UInt32.(eachindex(subset)))
```
`UInt32.(eachindex(subset))` materializes the whole vector which you definitely don't want to...

kalmarek committed:
remove overriding defs of copy/basis/Basis

kalmarek committed:
remove overriding defs of copy/basis/Basis

kalmarek committed:
remove action_on_group in favour of SW.action

kalmarek committed:
remove action_on_group in favour of SW.action

kalmarek committed:
simplify defn of WedderburnAction

Use CachedExtensionHomomorphism when applicable to cache
permutation actions on bases.

kalmarek committed:
simplify defn of WedderburnAction

Use CachedExtensionHomomorphism when applicable to cache
permutation actions on bases.

kalmarek committed:
rename fields of PSDBasisElement

kalmarek committed:
rename fields of PSDBasisElement

kalmarek committed:
rewrite action "on integers" via CartesianIndices

kalmarek committed:
rewrite action "on integers" via CartesianIndices

kalmarek committed:
WIP: encapsulate integers in LinIdx

This way we can use UInt32s as elements of the basis

kalmarek committed:
WIP: encapsulate integers in LinIdx

This way we can use UInt32s as elements of the basis

piotrmizerka committed:
we don't need quotient-slash definition

piotrmizerka committed:
we don't need quotient-slash definition

piotrmizerka committed:
rewrite action "on integers" via CartInd completed

piotrmizerka committed:
rewrite action "on integers" via CartInd completed

piotrmizerka committed:
keep manifest with url to PropertyT

piotrmizerka committed:
keep manifest with url to PropertyT

kalmarek commented:
@piotrmizerka use `git rebase` to move this branch over the master!

piotrmizerka commented:
> @piotrmizerka use `git rebase` to move this branch over the master!

which branch: Wedderburn or the one for our arxiv article?

kalmarek commented:
@piotrmizerka move this branch on top of master/main

piotrmizerka committed:
Merge branch 'Wedderburn' of https://github.com/piotrmizerka/LowCohomologySOS into Wedderburn

piotrmizerka committed:
an info what to do about induction from sl3 to sl4

piotrmizerka committed:
adjust coeffs in two matrices to another

piotrmizerka committed:
embeddings of SL(n,Z) into SL(m,Z)

piotrmizerka committed:
we make use now of upper and lower Laplacian

piotrmizerka committed:
symmetrization of matrices over group rings of sln

piotrmizerka committed:
upper and lower Laplacian: small correction

piotrmizerka committed:
symmetrized induction from sl3 to sl4

piotrmizerka committed:
a little tidy up in sln induction scripts

piotrmizerka committed:
temporary debug commands in sln induction

piotrmizerka committed:
embeddings moved to a separate file

piotrmizerka committed:
Laplacians moved to a separate file

piotrmizerka committed:
@time without descriptions as it caused errors

piotrmizerka committed:
Laplacians moved to a separate file

two functions yet to moved. Different tiny changes made

piotrmizerka committed:
PropertyT defines already ^ for words

piotrmizerka committed:
sln embedding tests

piotrmizerka committed:
separate files for Laplacian tests prepared

piotrmizerka committed:
unnecessary code removed

piotrmizerka committed:
symmetrization stuff put into one file

piotrmizerka committed:
induction script hopefully more readible

embedding and symmetrization stuff as package elements now

piotrmizerka committed:
description commit for action on real matrices

piotrmizerka committed:
do we want to Wedderburn-symmetrize sl3?

piotrmizerka committed:
more appropriate comment in embed_matrix

piotrmizerka committed:
wreath action on sln

piotrmizerka committed:
sln induction: playing with residual cetification

piotrmizerka committed:
sln Laplacians reparametrized

piotrmizerka committed:
sln induction: set cpu threads

piotrmizerka committed:
residual check on bigger radius

piotrmizerka committed:
enlarge int type for SAutF4 computation

piotrmizerka committed:
keep tghe whole relation set for sln

piotrmizerka committed:
add one parameter in adjustment

piotrmizerka committed:
extend sos problem to allow additional summands

piotrmizerka committed:
try removal of commutators with distinct indices

piotrmizerka committed:
don't print preprocessing info in induction script

piotrmizerka committed:
don't print auxiliary info

piotrmizerka committed:
it seems to be a relationship for upper Laplacians

piotrmizerka committed:
tiny changes

piotrmizerka committed:
Merge branch 'main' into Wedderburn

piotrmizerka committed:
we don't need action parameter for Wedderburn data

piotrmizerka committed:
Laplacians for sln and sautfn in a polymorphic way

piotrmizerka committed:
tests for relations in slns and sautfns

piotrmizerka committed:
wreath action on sln corrected

piotrmizerka committed:
further corrections of wreath action for sln

piotrmizerka committed:
starate relations for wreath action for sln

piotrmizerka committed:
Wedderburn approximation tests prettied up

piotrmizerka committed:
scaling for wreath products hopefully fixed

piotrmizerka committed:
fix types in action on matrix over RG

piotrmizerka committed:
wip: action formulae have to be studied

Although, it seems to be working...

piotrmizerka committed:
wip: decide which relations to take and cp their #

piotrmizerka committed:
check invariance of matrix to be decomposed

piotrmizerka committed:
test that star conj kills first upper Laplacian

piotrmizerka committed:
spectral_gap_elements: correct  arg indication

piotrmizerka committed:
tiny changes

piotrmizerka committed:
Adj symmetrization

Looks that Adj symmetrizes well!

piotrmizerka committed:
tiny changes

piotrmizerka committed:
decompose sq into diag and non-diag part

piotrmizerka committed:
there is no profit from wreath sym for sl3

piotrmizerka committed:
rfeturn to the single Sq

piotrmizerka committed:
remove unnecessary commit

piotrmizerka committed:
we don't need subset perm function

piotrmizerka committed:
test that Sq is SOS supported on hr  =1

piotrmizerka committed:
Merge pull request #17 from piotrmizerka/sln_wedderburn

Sln wedderburn

piotrmizerka committed:
split relations into op and adj for sln

piotrmizerka committed:
upper Laplacian for op rels plus Op is SOS!

piotrmizerka committed:
leave only the necessary stuff for sln induction

piotrmizerka committed:
sautf4 script customized for adj

piotrmizerka committed:
sl3 Laplacian plus adj certification

piotrmizerka committed:
split relations into sq, adj and op

piotrmizerka committed:
correct the test for wreath sln rels

piotrmizerka committed:
tiny descruption changes in Laplacian tests

piotrmizerka committed:
correct issues with the bound for lambda

piotrmizerka committed:
remove unnecessary sos_problem function

piotrmizerka committed:
adj part for sl3 without Wedderburn

piotrmizerka committed:
keep the original computation for Delta_1 for sl3z

piotrmizerka committed:
the whole computation for adj for sl3z

piotrmizerka committed:
sl3_adj from prec sol (todo: unable to push sol!!)

piotrmizerka committed:
another try to push an .sjl file

piotrmizerka committed:
activate package in scripts

piotrmizerka committed:
Merge pull request #15 from piotrmizerka/Laplacian_induction

induction from sl3 to sl4
